### PR TITLE
test: add the loadtime report tool

### PR DESCRIPTION
**Pull request by williambanfield**
This pull request adds the `report` tool and modifies the `loadtime` libraries to better support its use.
It moves the previously named `loadtime` tool to be called just `load`, built under `test/loadtime/cmd/load/`. The new report tool is built from `test/loadtime/cmd/report`. These commands cannot easily be made subcommands in the same binary because `load` makes use of the `tm-load-test` framework, which provides flags and subcommands of its own. Making both subcommands of the same binary would require surgery to that code base and likely is not worth it.

The `report` tool works by opening a handle to the Tendermint blockstore and reading all of the transactions out of the block. It calculates the latency of each transaction by differencing the timestamp listed in the transaction `Payload` and the timestamp of the block. It uses these latencies to calculate a set of metrics including the minimum, maximum, average and standard deviation value.

The tool currently treats all of the transactions it finds as being part of the _same_ experimental run, but minor additional work is needed to separate transactions by experimental variables (Connections / Size / Rate as listed in the `Payload` data). See: https://github.com/tendermint/tendermint/issues/9352

Sample Output

```
./build/report --database-type goleveldb --data-dir ~/.tendermint/data
Total Valid Tx: 9000
Total Invalid Tx: 0
Minimum Latency: 351.103223ms
Maximum Latency: 1.040886707s
Average Latency: 703.895594ms
Standard Deviation: 214.400818ms
```

closes: #9331 

#### PR checklist

- [ ] Tests written/updated, or no tests needed
- [ ] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [ ] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

File: test/loadtime/report/report.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	txc := make(chan txData)
+	pdc := make(chan payloadData, poolSize)
+
+	wg := &sync.WaitGroup{}
+	wg.Add(poolSize)
+	for i := 0; i < poolSize; i++ {
+		go func() {
+			defer wg.Done()
+			for b := range txc {
+				p, err := payload.FromBytes(b.tx)
```
Comment (williambanfield) [inline test/loadtime/report/report.go:60]: Parsing is slow but does not require coordination so it is done it a go routine here.

File: test/loadtime/README.md [hunk 2]
```diff
@@ -1,32 +1,66 @@
...
     --endpoints ws://localhost:26657/websocket
 ```
+
+## `report`
+
+The `report` binary is built when `make build` is invoked. The `report` tool
+reads all of the blocks from the specified blockstore database and calculates
+transaction latency metrics. `report` reads transactions generated by `load`
+and uses the difference between the timestamp contained in the transaction and
+the timestamp of the block the transaction was executed in determine transaction latency.
```
Comment (sergio-mena) [inline test/loadtime/README.md:45]:
```suggestion
the timestamp of the block the transaction was executed in to determine transaction latency.
```

File: test/loadtime/cmd/report/main.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	"time"
+
+	"github.com/tendermint/tendermint/store"
+	"github.com/tendermint/tendermint/test/loadtime/report"
+	dbm "github.com/tendermint/tm-db"
+)
+
+var (
+	db     = flag.String("database-type", "goleveldb", "the type of database holding the blockstore")
+	dir    = flag.String("data-dir", "~/.tendermint/data", "path to the directory containing the tendermint databases")
```
Comment (sergio-mena) [inline test/loadtime/cmd/report/main.go:20]: Consider reading TMHOME when constructing the default value of "data-dir" (for consistency with Tendermint).

File: test/loadtime/cmd/report/main.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+		w := csv.NewWriter(cf)
+		err = w.WriteAll(toRecords(r.All))
+		if err != nil {
+			panic(err)
+		}
+		return
+	}
+
+	fmt.Printf(""+
+		"Total Valid Tx: %d\n"+
```
Comment (sergio-mena) [inline test/loadtime/cmd/report/main.go:69]: I would also print the sample's size, i.e., now many transactions (and blocks?) were processed

Comment (sergio-mena) [reply test/loadtime/cmd/report/main.go:69]: Oops! Forget this comment. Not enough coffee! Going to fix that now :-)

File: test/loadtime/report/report.go (hunk 1 above)
Comment (sergio-mena) [reply test/loadtime/report/report.go:60]: I appreciate the effort done to parallelize transaction processing.
The (minor) problem I see is that some latencies will come out unordered. So if we get the results and want to plot them, we'll loose information.
Anyway, in that case we can set `poolSize` to 1. But would be good to be able to do it via a flag (--no-parallel, or something like that)

File: test/loadtime/report/report.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+				if err != nil {
+					pdc <- payloadData{err: err}
+					continue
+				}
+				l := b.bt.Sub(p.Time.AsTime())
```
Comment (sergio-mena) [inline test/loadtime/report/report.go:65]: We agreed that, for this scheme to work well, we'll need to configure NTP on nodes involved.
In order to (try to) catch the fact that clocks are not synchronized, could you add some (rough) sanity checks here? For instance:
* l should be positive
* I should be less than, say, 120 seconds

These could also be done by inspecting the results provided by the report (say, the script that calls `report`). Not sure what's best here.

**Review by sergio-mena — APPROVED** [review]
Cool job!

File: test/loadtime/report/report.go (hunk 1 above)
Comment (williambanfield) [reply test/loadtime/report/report.go:60]: > The (minor) problem I see is that some latencies will come out unordered. So if we get the results and want to plot them, we'll loose information.

This is true, but not a requirement of the tool at the moment. We can certainly change it to be ordered by adding a height / index field to the metadata given to the parallel workers and then performing a final sort of the data at the end (or heap sorting during the creation). Either sort will certainly be faster than serial parsing.

File: test/loadtime/cmd/report/main.go (hunk 3 above)
Comment (williambanfield) [reply test/loadtime/cmd/report/main.go:20]: TBH, I'd almost rather have no default for this flag. The reason I've used a flag here instead of using TMHOME is that the tool is, in the current iteration, unlikely to be used on the machine that the db was generated on. The thinking is that we'll zip up the `/data` directory and pull it down for analysis. If that's the targeted usecase, it doesn't really make sense to use the `TMHOME` since it's unlikely we'll actually be using the normal Tendermint data directory on a machine and doing so may accidentally rip data out of a test instance on a users's machine when they don't intend to.

File: test/loadtime/report/report.go (hunk 5 above)
Comment (williambanfield) [reply test/loadtime/report/report.go:65]: > the resulting duration should be positive

I think it may make sense to just report a count of negative durations. I doubt we want to error in those cases since the data is still perhaps interesting. Does this seem reasonable?

File: test/loadtime/cmd/report/main.go (hunk 3 above)
Comment (sergio-mena) [reply test/loadtime/cmd/report/main.go:20]: Yeah, I like the "no default" solution.

File: test/loadtime/report/report.go (hunk 5 above)
Comment (sergio-mena) [reply test/loadtime/report/report.go:65]: Thinking a bit more about this, probably a single flag in the report (set to true if _any_ delta is negative or ridiculously big) should suffice.
If we have evidence the the clocks aren't synchronized (because we forgot, or NTP daemon crashed, or was accidentally stopped in one of the machines), the simplest course of action to me would be to repeat the test from scratch with a good NTP setup, rather than trying to extract info from a report with skewed measurements.

In short, a count of negative (or very big) durations would be good, but we can also have just a general flag ("clocks likely not in sync").

File: test/loadtime/report/report.go (hunk 1 above)
Comment (sergio-mena) [reply test/loadtime/report/report.go:60]: I see. Could add a one-line comment making clear the latencies we output are unordered?

File: test/loadtime/report/report.go (hunk 5 above)
Comment (williambanfield) [reply test/loadtime/report/report.go:80]: > ridiculously big

We have the `Min` and `Max` so users can determine this on their own. 

> but we can also have just a general flag ("clocks likely not in sync").

I'd prefer to not try to be much smarter than we need to be at the possible expense of clarity. Adding such a flag is doing the work of interpreting the data for the user which I don't think this tool should be doing. IMO, it should report the data in a manner that helps the user draw conclusions about what happened during the experimental run. It _should_ be clear enough from the outputs given (large number of negative durations, or huge values for Max and Min) that the clocks are not in sync.
